ethcore/src/state_db.rs

```diff
@@ -221,11 +262,31 @@ impl StateDB {
 	/// or if the account is not cached.
 	pub fn get_cached<F, U>(&self, a: &Address, f: F) -> Option<U>
 		where F: FnOnce(Option<&mut Account>) -> U {
-		if !self.is_canon {
+		let mut cache = self.account_cache.lock();
+		if !Self::is_allowed(a, &self.parent_hash, &cache.modifications) {
 			return None;
 		}
-		let mut cache = self.account_cache.lock();
 		cache.accounts.get_mut(a).map(|c| f(c.as_mut()))
 	}
+
+	fn is_allowed(addr: &Address, parent_hash: &Option<H256>, modifications: &VecDeque<(H256, HashSet<Address>)>) -> bool {
+		if let &Some(ref parent) = parent_hash {
```

Comment:
block with complex logic inside can be replaced with opposite condition and return in the beginning reducing nesting
```
let parent = match parent_hash { None => return false, Some(ref parent) => parent };
```